cpp/src/arrow/python/io.cc

```diff
@@ -40,20 +40,35 @@ namespace py {
 // calling any methods
 class PythonFile {
  public:
-  explicit PythonFile(PyObject* file) : file_(file) { Py_INCREF(file_); }
+  explicit PythonFile(PyObject* file) : file_(file) { Py_INCREF(file); }
 
-  ~PythonFile() { Py_DECREF(file_); }
+  Status CheckClosed() const {
+    if (!file_) {
+      return Status::Invalid("operation on closed Python file");
+    }
+    return Status::OK();
+  }
 
   Status Close() {
-    // whence: 0 for relative to start of file, 2 for end of file
-    PyObject* result = cpp_PyObject_CallMethod(file_, "close", "()");
-    Py_XDECREF(result);
-    PY_RETURN_IF_ERROR(StatusCode::IOError);
+    if (file_) {
+      PyObject* result = cpp_PyObject_CallMethod(file_.obj(), "close", "()");
+      Py_XDECREF(result);
+      PY_RETURN_IF_ERROR(StatusCode::IOError);
```

Comment:
For safety, I'd `file_.reset()` before returning the error so subsequent calls returns "closed = true".

Reply:
Good point!